channels/rdpdr/server/rdpdr_main.c

```diff
@@ -591,37 +825,968 @@ static int rdpdr_server_stop(RdpdrServerContext* context)
+
+	if (ioStatus != STATUS_SUCCESS)
+	{
+		/* Invoke the rename file completion routine. */
+		context->OnDriveRenameFileComplete(context, irp->CallbackData, ioStatus);
+
+		/* Destroy the IRP. */
+		rdpdr_server_irp_free(irp);
+
+		return;
+	}
+
+	Stream_Read_UINT32(s, fileId); /* FileId (4 bytes) */
+	Stream_Read_UINT8(s, information); /* Information (1 byte) */
+
+	printf("%s: FileId=%u\n", __FUNCTION__, fileId); fflush(stdout);
+
+	/* Setup the IRP. */
+	irp->CompletionId = context->priv->NextCompletionId++;
+	irp->Callback = rdpdr_server_drive_rename_file_callback2;
+	irp->DeviceId = deviceId;
+	irp->FileId = fileId;
+
+	rdpdr_server_enqueue_irp(context, irp);
+
+	/* Send a request to rename the file */
+	rdpdr_server_send_device_file_rename_request(context, deviceId, fileId, irp->CompletionId, irp->ExtraBuffer);
+}
+
+static BOOL rdpdr_server_drive_rename_file(RdpdrServerContext* context, void* callbackData, UINT32 deviceId, const char* oldPath, const char* newPath)
+{
+	RDPDR_IRP* irp;
+
+	/* Create an IRP. */
+	irp = rdpdr_server_irp_new();
+	if (irp == NULL) return FALSE;
+
+	irp->CompletionId = context->priv->NextCompletionId++;
+	irp->Callback = rdpdr_server_drive_rename_file_callback1;
+	irp->CallbackData = callbackData;
+	irp->DeviceId = deviceId;
+	strncpy(irp->PathName, oldPath, sizeof(irp->PathName));
+	strncpy(irp->ExtraBuffer, newPath, sizeof(irp->ExtraBuffer));
+
+	rdpdr_server_convert_slashes(irp->PathName, sizeof(irp->PathName));
+	rdpdr_server_convert_slashes(irp->ExtraBuffer, sizeof(irp->ExtraBuffer));
+
+	printf("PathName=%s\n", irp->PathName); fflush(stdout);
+	printf("ExtraBuffer=%s\n", irp->ExtraBuffer); fflush(stdout);
+
+	rdpdr_server_enqueue_irp(context, irp);
+
+	/* Send a request to open the file. */
+	rdpdr_server_send_device_create_request(
+		context, deviceId, irp->CompletionId, irp->PathName,
+		FILE_READ_DATA | SYNCHRONIZE,
+		FILE_SYNCHRONOUS_IO_NONALERT,
+		FILE_OPEN);
+
+	return TRUE;
+}
+
+RdpdrServerContext* rdpdr_server_context_new(HANDLE vcm)
+{
+	RdpdrServerContext* context;
+	context = (RdpdrServerContext*) malloc(sizeof(RdpdrServerContext));
+
+	if (context)
+	{
+		ZeroMemory(context, sizeof(RdpdrServerContext));
+
+		context->vcm = vcm;
+
+		context->Start = rdpdr_server_start;
+		context->Stop = rdpdr_server_stop;
+
+		context->DriveCreateDirectory = rdpdr_server_drive_create_directory;
+		context->DriveDeleteDirectory = rdpdr_server_drive_delete_directory;
+		context->DriveQueryDirectory = rdpdr_server_drive_query_directory;
+		context->DriveOpenFile = rdpdr_server_drive_open_file;
+		context->DriveReadFile = rdpdr_server_drive_read_file;
+		context->DriveWriteFile = rdpdr_server_drive_write_file;
+		context->DriveCloseFile = rdpdr_server_drive_close_file;
+		context->DriveDeleteFile = rdpdr_server_drive_delete_file;
+		context->DriveRenameFile = rdpdr_server_drive_rename_file;
+
+		context->priv = (RdpdrServerPrivate*) malloc(sizeof(RdpdrServerPrivate));
```

Comment:
I would assume that context->priv is required for later use. Therefore it would make sense to free the context and return NULL if the above malloc fails.